flow/BlobCipher.cpp

```diff
@@ -116,74 +116,60 @@ void BlobCipherKey::reset() {
 // BlobKeyIdCache class methods
 
 BlobCipherKeyIdCache::BlobCipherKeyIdCache()
-  : domainId(ENCRYPT_INVALID_DOMAIN_ID), latestBaseCipherKeyId(ENCRYPT_INVALID_CIPHER_KEY_ID),
-    latestRandomSalt(ENCRYPT_INVALID_RANDOM_SALT) {}
+  : domainId(ENCRYPT_INVALID_DOMAIN_ID), latestBaseCipherKeyId(), latestRandomSalt() {}
 
 BlobCipherKeyIdCache::BlobCipherKeyIdCache(EncryptCipherDomainId dId)
-  : domainId(dId), latestBaseCipherKeyId(ENCRYPT_INVALID_CIPHER_KEY_ID), latestRandomSalt(ENCRYPT_INVALID_RANDOM_SALT) {
+  : domainId(dId), latestBaseCipherKeyId(), latestRandomSalt() {
 	TraceEvent("Init_BlobCipherKeyIdCache").detail("DomainId", domainId);
 }
 
 BlobCipherKeyIdCacheKey BlobCipherKeyIdCache::getCacheKey(const EncryptCipherBaseKeyId& baseCipherKeyId,
                                                           const EncryptCipherRandomSalt& salt) {
-	if (baseCipherKeyId == ENCRYPT_INVALID_CIPHER_KEY_ID || salt == ENCRYPT_INVALID_RANDOM_SALT) {
+	if (baseCipherKeyId == ENCRYPT_INVALID_CIPHER_KEY_ID) {
```

Comment:
any reason we removed the 'salt' validity check? We can force the invariant that salt > 0?

Reply:
I changed to use Optional for latest salt so there's no need to add an invalid salt value as a placeholder. plus... I'm a bit paranoid as we don't have control when we create the random salt in theory it can be 0, so 0 is still a valid salt. True we can control how the salt is generated, but there's probably no need to do so.

Reply:
> as when we create the random salt in theory it can be 0

RandomSalt generation is controlled by BlobCipher and happens as part of BaseCipher KeyDerivation (https://github.com/apple/foundationdb/blob/main/flow/BlobCipher.cpp#L48); if it fair that `std::mt19937 random;` can generate `0` as the output, we should add `1` to the BlobCipherKey::BlobCipherKey() to distinguish `salt` from its default/invalid value i.e. `0`

However, once we make above changes, the library controls the `salt` generation and I don't think it can be modified other than tampering with plaintext EncryptHeader (brute force attack..)

Reply:
> changed to use Optional for latest salt 

Am not sure if Optional is the right choice, the values will be always present after a first latestKey gets inserted; to me they is a "default" value at the time of object construction and then value keeps updating with time. 

It is not a biggy either ways and we have discussed this more than needed, if you prefer Optional, please proceed with it; no strong opinions from my side. 
Thanks .

Reply:
The "default" value effectively make the salt an Optional. To me, Optional is just an explicit way to represent nullable value.

And I put back the invalid random salt check, just in case you think it is handy somewhere else.